File: v1/examples/python/01_login.py
"""
01_login.py - TS-API v1 Authentication Methods

Three authentication methods:
  1. Session Login   - POST /api/v1/auth/login (JSON body, cookie-based)
  2. JWT Token       - POST /api/v1/auth/login (JSON body, token-based)
  3. API Key         - X-API-Key header (v1 endpoints only)

NOTE: JWT tokens only work on /api/v1/auth/* endpoints.
      For data endpoints (channels, events, etc.), use Legacy Session or API Key.
NOTE: API Key authentication is supported on v1 endpoints only.
      v0 endpoints (/api/*) reject X-API-Key with 401 Unauthorized.
"""

import os
import sys
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config import NVR_USER, NVR_PASS, BASE_URL
from http_client import NvrClient


def demo_session_login():
    """Method 1: Session Login (POST with JSON body).

    POST /api/v1/auth/login with {"auth": "base64(username:password)"}
    The server sets a session cookie on successful login.
    All subsequent requests carry the cookie automatically.
    """
    print("=== Session Login ===")
    client = NvrClient()

    # Login via POST (credentials in JSON body, not URL)
    ok = client.login(NVR_USER, NVR_PASS)
    print(f"Login: {'OK' if ok else 'FAILED'}")

    if ok:
        # Session cookie is now set; data endpoints work
        r = client.get('/api/v1/channel')
        print(f"GET /api/v1/channel -> {r.status_code}")

        # Always logout to release the session
        client.logout()
        print("Logged out")


def demo_jwt():
    """Method 2: JWT Token (accessToken + refreshToken).

    POST /api/v1/auth/login -> {accessToken, refreshToken, expiresIn, tokenType}
    POST /api/v1/auth/refresh -> {accessToken, expiresIn, tokenType}
    POST /api/v1/auth/logout  (revoke refreshToken)
    """
    print("\n=== JWT Login ===")
    client = NvrClient()

    # 1) Login -> get accessToken + refreshToken
    tokens = client.jwt_login(NVR_USER, NVR_PASS)
    if not tokens:
        print("JWT login failed")
        return

    access_token = tokens['accessToken']
    refresh_token = tokens['refreshToken']
    print(f"Access Token:  {access_token[:30]}...")
    print(f"Refresh Token: {refresh_token[:30]}...")
    print(f"Expires In:    {tokens['expiresIn']}s")

    # 2) Use accessToken for auth endpoints
    r = client.get('/api/v1/auth/apikey',
                    headers={'Authorization': f'Bearer {access_token}'})
    print(f"GET /api/v1/auth/apikey -> {r.status_code}")

    # 3) Refresh accessToken using refreshToken
    new_tokens = client.jwt_refresh(refresh_token)
    if new_tokens:
        print(f"Refreshed:     {new_tokens['accessToken'][:30]}...")

    # 4) Logout (revoke refreshToken)
    client.jwt_logout(refresh_token)
    print("JWT logged out (refreshToken revoked)")


def demo_api_key():
    """Method 3: API Key (Create -> Use -> List -> Delete).

    POST /api/v1/auth/apikey        (Create, admin JWT required)
    X-API-Key header                (Use, v1 endpoints only)
    GET /api/v1/auth/apikey         (List)
    DELETE /api/v1/auth/apikey/{id} (Delete)

    NOTE: API Key only works on v1 endpoints (/api/v1/*).
          v0 endpoints (/api/*) reject API Key with 401.
    """
    print("\n=== API Key ===")
    client = NvrClient()

    # 1) Obtain admin token via JWT login
    token_data = client.jwt_login(NVR_USER, NVR_PASS)
    if not token_data:
        print("JWT login failed")
        return
    token = token_data['accessToken']
    auth = {'Authorization': f'Bearer {token}'}

    # 2) Create API Key
    r = client.post('/api/v1/auth/apikey', json={
        'name': 'example-integration',
        # 'permissions': ['remote'],       # Optional: specify permissions (default: remote)
        # 'channels': [1, 2],              # Optional: restrict accessible channels
        # 'ipWhitelist': ['192.168.0.0/24'], # Optional: IP restriction
    }, headers=auth)
    print(f"Create API Key: {r.status_code}")
    if r.status_code != 200:
        return

    data = r.json()
    key_id = data['id']
    api_key = data['key']
    print(f"  Key ID: {key_id}")
    print(f"  API Key: {api_key[:24]}...")
    if 'message' in data:
        print(f"  WARNING: {data['message']}")

    # 3) Access data endpoints using API Key (no login required)
    r = client.get('/api/v1/channel', headers={'X-API-Key': api_key})
    print(f"Use API Key -> GET /api/v1/channel: {r.status_code}")

    # 4) List API Keys
    r = client.get('/api/v1/auth/apikey', headers=auth)
    print(f"List API Keys: {r.status_code} ({len(r.json())} keys)")

    # 5) Delete API Key
    r = client.delete(f'/api/v1/auth/apikey/{key_id}', headers=auth)
    print(f"Delete API Key: {r.status_code}")


if __name__ == '__main__':
    demo_session_login()
    demo_jwt()
    demo_api_key()

File: v1/examples/csharp/01_Login.cs
// 01_Login.cs - v1 API Authentication Example
//
// v1 API supports two authentication methods:
//
// 1. JWT auth (POST with JSON body)
//    POST /api/v1/auth/login  {"auth":"base64(username:password)"}
//    POST /api/v1/auth/logout {"refreshToken":"..."}
//
// 2. JWT auth (token-based)
//    POST /api/v1/auth/login -> {accessToken, refreshToken, expiresIn, tokenType}
//
// 3. API Key (external integration)
//    POST /api/v1/auth/apikey (create, requires admin JWT)
//    X-API-Key header (use, v1 endpoints only)
//
// IMPORTANT: JWT only works on /api/v1/auth/* endpoints!
// API Key authentication is supported on v1 endpoints only. v0 returns 401.

using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace TsApiExamples.V1
{
    class Login
    {
        static async Task Main(string[] args)
        {
            using var client = new NvrClient();
            Console.WriteLine($"NVR: {client.BaseUrl}");

            // ── Method 1: Session auth ──
            Console.WriteLine("\n=== Session Login ===");
            var ok = await client.LoginAsync();
            Console.WriteLine($"Login: {(ok ? "OK" : "FAILED")}");

            if (ok)
            {
                var r = await client.GetAsync("/api/v1/channel");
                Console.WriteLine($"GET /api/v1/channel -> {(int)r.StatusCode}");

                await client.LogoutAsync();
                Console.WriteLine("Logged out");
            }

            // ── Method 2: JWT auth (accessToken + refreshToken) ──
            Console.WriteLine("\n=== JWT Login ===");
            using var jwtDoc = await client.JwtLoginAsync();
            if (jwtDoc == null)
            {
                Console.WriteLine("JWT login failed");
                return;
            }

            var jwtRoot = jwtDoc.RootElement;
            var accessToken = jwtRoot.GetProperty("accessToken").GetString();
            var refreshToken = jwtRoot.GetProperty("refreshToken").GetString();
            Console.WriteLine($"Access Token:  {accessToken?.Substring(0, 20)}...");
            Console.WriteLine($"Refresh Token: {refreshToken?.Substring(0, 20)}...");
            Console.WriteLine($"Expires In:    {jwtRoot.GetProperty("expiresIn").GetInt32()}s");

            // Refresh accessToken using refreshToken
            using var refreshDoc = await client.JwtRefreshAsync(refreshToken);
            if (refreshDoc != null)
            {
                var newToken = refreshDoc.RootElement.GetProperty("accessToken").GetString();
                Console.WriteLine($"Refreshed:     {newToken?.Substring(0, 20)}...");
                accessToken = newToken;  // use new token
            }

            // JWT logout (revoke refreshToken)
            // await client.JwtLogoutAsync(refreshToken);  // skip: need token for API Key demo below

            // ── Method 3: API Key lifecycle (Create -> Use -> List -> Delete) ──
            Console.WriteLine("\n=== API Key ===");
            var bearer = new AuthenticationHeaderValue("Bearer", accessToken);

            // 1) Create API Key
            var createReq = new HttpRequestMessage(HttpMethod.Post, "/api/v1/auth/apikey");
            createReq.Headers.Authorization = bearer;
            createReq.Content = new StringContent(
                JsonSerializer.Serialize(new { name = "example-integration" }),
                Encoding.UTF8, "application/json");
            var createRes = await client.SendAsync(createReq);
            Console.WriteLine($"Create API Key: {(int)createRes.StatusCode}");

            if (!createRes.IsSuccessStatusCode) return;

            using var keyDoc = JsonDocument.Parse(await createRes.Content.ReadAsStringAsync());
            var keyRoot = keyDoc.RootElement;
            var keyId = keyRoot.GetProperty("id").GetString();
            var apiKey = keyRoot.GetProperty("key").GetString();
            Console.WriteLine($"  Key ID: {keyId}");
            Console.WriteLine($"  API Key: {apiKey?.Substring(0, Math.Min(24, apiKey?.Length ?? 0))}...");

            // 2) Use API Key (no login needed)
            var useReq = new HttpRequestMessage(HttpMethod.Get, "/api/v1/channel");
            useReq.Headers.Add("X-API-Key", apiKey);
            var useRes = await client.SendAsync(useReq);
            Console.WriteLine($"Use API Key -> GET /api/v1/channel: {(int)useRes.StatusCode}");

            // 3) List API Keys
            var listReq = new HttpRequestMessage(HttpMethod.Get, "/api/v1/auth/apikey");
            listReq.Headers.Authorization = bearer;
            var listRes = await client.SendAsync(listReq);
            Console.WriteLine($"List API Keys: {(int)listRes.StatusCode}");

            // 4) Delete API Key
            var delReq = new HttpRequestMessage(HttpMethod.Delete, $"/api/v1/auth/apikey/{keyId}");
            delReq.Headers.Authorization = bearer;
            var delRes = await client.SendAsync(delReq);
            Console.WriteLine($"Delete API Key: {(int)delRes.StatusCode}");
        }
    }
}
